### ISSUE TITLE
Utilize pytest cache for a faster observable TTS

### ISSUE THREAD

**Issue by huydhn** [pytorch/pytorch#84183]
### 🚀 The feature, motivation and pitch

Now that PyTorch has started using pytest, we should explore [pytest cache](https://docs.pytest.org/en/latest/how-to/cache.html) for a faster observable TTS.  Specifically, we should consider using `--failed-first` and `--new-first` options when running pytest so that when devs re-run the workflow (push a new commit to the PR, rebase, retry, revert a commit), failed tests from the previous run should be executed first. This has several benefits:

* When a new commit is pushed to the PR to fix broken tests, devs only need to wait a shorter time to confirm that these tests have been fixed if they are run first.  This is especially useful to re-run failed tests for platforms that devs don't have access to and rely on the CI to iterate on a fix
* When a change is reverted for breaking trunk. Re-running failed tests first helps confirm that the issue has been fixed
* We can even consider to use `--last-failed` to re-run only flaky tests when there is no change (retry workflow)

Pytest cache from the previous run will need to be persisted on S3 for the current test run to pick it up. This is similar to pytest-report artifact.

Some limitations that I can think of:
* Are all the unit tests independent? They should but there are so many of them so we need to try it out to see if something weird happens when the test order changes
* There doesn't seem to be an equivalent option for unittest. This is a pytest-only feature

cc @seemethere   @malfet @pytorch/pytorch-dev-infra @clee2000 

### Additional context

PyTorch CI optimization

**Comment by suo** [comment]
I think this is a good idea—but very few of our tests run in pytest, so the impact is limited. Perhaps we should invest in a project to migrate more of our testing to pytest (it seems there are lots of things we want in pytest-land)

**Comment by huydhn** [comment]
> I think this is a good idea—but very few of our tests run in pytest, so the impact is limited. Perhaps we should invest in a project to migrate more of our testing to pytest (it seems there are lots of things we want in pytest-land)

Yeah, I kind of rely on @clee2000 test parallelization project here, in which tests are slowly moving to run in parallel with pytests.  AFAIK, only large test files like test_ops are migrated so far, but it would be nice to have the remaining tests moving over as well.

**Comment by huydhn** [comment]
cc @ZainRizvi as there is a recent surge of interests in having this feature.

**Comment by ZainRizvi** [comment]
We now first run newly edited tests and tests that failed in any previous iteration of a PR's CI as of  https://github.com/pytorch/pytorch/pull/101123
